Initial project setup

Ipiano opened the pull request:
* Code formatting rules
* Pull and build dependencies
* Build dummy gtest

Draft until github actions can be added

Ipiano committed:
Add formatting and linting configuration

Ipiano committed:
Add initial cmake project

Ipiano committed:
Add googletest dependency

By default, this will use the FindGTest.cmake file included in the
cmake folder. This file attempts to clone googletest during the cmake
configure step and then add it with add_subdirectory.

Since a replacement find_package implementation could just use the
GTest config cmake files from a googletest installation, this will
alias the four targets added by the project to match what would be
added by the installed config files.

Ipiano committed:
Add shapelib dependency

This will find and add shapelib similarly to how we are finding and
adding googletest with FetchContent.

By default, shapelib will attempt to fetch test data so that it can be
used for tests. This did not work out of the box, and we were unable
to determine how to make it work; so we opted to disable the testing
from shapelib altogether.

The shapelib cmake files do not provide a method to disable testing; it
looks like you should be able to use -DBUILD_TEST=OFF to disable them,
but because the variable is not a CACHE or option() variable, this
doesn't actually work. In order to disable the tests, we are applying
a patch step after fetching the library which will use a cmake
sub-invocation to replace set(BUILD_TEST ON) with set(BUILD_TEST OFF)
throughout the project's cmake file.

Ipiano committed:
Determine if being built as standalone project

Identifying this allows us to default build variables or exclude generic
targets more appropriately when this library is included as a
sub-project.

Ipiano committed:
Configure project tests

Ipiano committed:
Link gtest with -isystem

Notgnoshi committed:
Add formatting workflow with clang-format

Notgnoshi committed:
Add cmake-format to formatting workflow

Notgnoshi committed:
Convert cmake-format config to json

The cmake-format-lint-action GitHub action does not have the python-yaml
package installed, so we either have to build our own action or convert
our configuration file to JSON. We opted for the later.

Notgnoshi committed:
Add workflow for linting project

Notgnoshi commented:
TODO:
* [x] Build library
* [x] Build and run tests in pipeline
* [x] Code coverage
* [x] Badges
* [x] Doxygen
* [x] GitHub pages for doxygen?

Longer term:
* [ ] Benchmarks
* [ ] Test data!

Ipiano committed:
Alias run-shpxx-tests to check if root project

Notgnoshi committed:
Add Doxygen build to CMake

The shpxx-build-docs target is added by default when shpxx is being
built as a top-level project (not being consumed by another project).
In this case, the project internal documentation is also build, but can
be disabled with the SHPXX_BUILD_PRIVATE_DOCS option. This is expected
to be used for public-facing documentation deployed e.g., to a GitHub
pages site for the project.

Notgnoshi committed:
Export a clang compilation database by default

This is immensely helpful for using tooling that provides language
server capabilities, like clangd or clang-tidy.

Notgnoshi committed:
Add Doxygen GitHub pages workflow

This workflow will build the documentation for PRs, and deploy the
documentation to https://notgnoshi.github.io/shpxx/ on merges to main.

Ipiano committed:
Add find_package file to find gcovr and gcov

Ipiano committed:
Add initial shpxx library target

Ipiano committed:
Add coverage report target

Notgnoshi committed:
Add badges for pass/fail workflows

Notgnoshi committed:
Use short workflow names for badges

The badges shown on the README use the workflow name, and long names
don't look good.

Ipiano committed:
Add workflow to build, test, and report coverage

Ipiano committed:
Add coverage percentage badge

Notgnoshi reviewed (approved):
+1 LGTM
